File: python/examples/hello.py
"""Zero-to-working VCP: build+sign a capability, run it end to end through the Gateway.

Run it from the ``python/`` directory:

    python examples/hello.py

Runs with the standard library alone — if ``cryptography`` is not installed the
SDK transparently falls back to a *labelled* HMAC signer, so the whole flow still
works (the signatures just announce they are a dev fallback, not real Ed25519).
Install the real thing with ``pip install vcp-python[crypto]``.

What this demonstrates, in order:
  1. Build a tiny capability manifest and sign it.
  2. Print its content-addressed ``capability_id`` (hash of the security contract).
  3. Hand it to the Gateway, which verifies it -> runs policy -> mints a single-use
     grant -> invokes an in-process provider -> verifies the provider's attestation.
"""

from __future__ import annotations

# --- Import resolution -------------------------------------------------------
# We want `python examples/hello.py` to work whether or not the package is
# installed. The unittest suite resolves imports by running from the package
# root (`-t .`); we mirror that by putting the `python/` dir (this file's parent
# directory) on sys.path so `vcp_sdk` / `vcp_gateway` import either way.
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from vcp_gateway import DefaultPolicy, Gateway, InMemoryProvider
from vcp_sdk import build_manifest, default_signer, propose_plan


def main() -> None:
    # The Gateway and the provider each own a signer. In VCP these are separate
    # trust domains: the Gateway signs grants/audit; the provider signs the
    # attestation over its result. We verify each against the *other* party's
    # public verifier below, so nobody is trusted on their own say-so.
    gateway_signer = default_signer()
    provider_signer = default_signer()

    # --- 1. Build + sign a tiny, read-only capability manifest ---------------
    # read-only means no external side effect and NO user approval, so this runs
    # unattended. The capability id is derived purely from the security-relevant
    # contract (issuer/name/version/schemas/effects/...), NOT from the summaries —
    # those are display strings excluded from identity (SPEC §4).
    manifest = build_manifest(
        issuer="did:web:example.com",        # who vouches for the contract
        provider="example.echo",             # who runs it
        name="echo.say",
        version="1.0.0",
        input_schema={
            "type": "object",
            "additionalProperties": False,   # strict: the Gateway rejects extra args
            "properties": {"text": {"type": "string"}},
            "required": ["text"],
        },
        output_schema={
            "type": "object",
            "properties": {"echoed": {"type": "string"}},
            "required": ["echoed"],
        },
        # read-only is the friction-free class: no side effect, no approval gate.
        effects={"class": "read-only", "external_side_effect": False},
        determinism={"class": "pure"},
        sandbox={"filesystem": "none", "network": [], "secrets": []},
        summary_for_user="Echo a short string back.",
        summary_for_model="Echo the given text. Read-only, no side effects.",
        signer=gateway_signer,               # sign the manifest so it can be verified
    )

    # --- 2. The content-addressed capability id ------------------------------
    # Same contract bytes anywhere in the world => same id. Tamper with any
    # security-relevant field and the id changes, so a swapped contract can't
    # masquerade as this one.
    capability_id = manifest["capability"]["id"]
    print("capability_id:", capability_id)

    # --- 3. The in-process provider ------------------------------------------
    # A real provider lives behind a network boundary; here it's a function. It
    # re-checks the grant audience and recomputes the argument hash before acting
    # (SPEC §8), then signs an attestation over its result so the Gateway can
    # prove the result really came from THIS capability on THESE arguments.
    provider = InMemoryProvider(
        capability_id,
        signer=provider_signer,
        handler=lambda args, dry_run: {"echoed": args["text"]},
    )

    # The plan_hash binds this call to an approved plan (SPEC §3.3). A grant is
    # only valid for the plan it was minted against, so a stolen grant can't be
    # replayed inside a different plan.
    arguments = {"text": "hello, vcp"}
    plan = propose_plan(
        [{"id": "s1", "capability": capability_id, "arguments": arguments, "effect": "read-only"}]
    )

    # --- 4. Drive the full Gateway pipeline ----------------------------------
    # verify manifest -> validate args (strict schema) -> policy decision ->
    # mint single-use proof-bound grant -> invoke provider -> verify attestation.
    # Any failure raises GatewayError and NO result reaches the caller (fail
    # closed, SPEC §19). We pass each party's verifier so trust is mutual.
    gateway = Gateway(
        policy=DefaultPolicy(),
        signer=gateway_signer,
        trusted_issuers={"did:web:example.com"},  # only accept contracts from this issuer
    )
    out = gateway.invoke(
        manifest=manifest,
        provider=provider,
        arguments=arguments,
        subject="user:alice",
        plan_hash=plan["plan_hash"],
        # The holder key thumbprint binds the grant to its holder (proof of
        # possession, SPEC §7); a leaked grant is useless without the matching key.
        holder_jkt="sha256:" + "0" * 64,
        manifest_verifier=gateway_signer.verifier(),
        attestation_verifier=provider_signer.verifier(),
    )

    # The result is labelled untrusted_tool_result: it carries taint so the
    # planner can't later let it silently authorize a privileged action (SPEC §12).
    print("result:        ", out["result"])
    print("grant_id:      ", out["grant_id"])
    print("attested hash: ", out["attestation"]["result_hash"])
    print("label:         ", out["label"])
    print("audit events:  ", [e["event"] for e in gateway.audit.events])


if __name__ == "__main__":
    main()

File: typescript/examples/hello.ts
// VCP in ~40 lines: build & sign a capability, then run it through the Gateway.
//
// VCP's core idea: the model/planner is NEVER an authority. A capability is a
// *signed, content-addressed contract*. An enforcing Gateway verifies that
// contract, asks policy, mints a single-use grant bound to the exact call, runs
// the provider, and refuses to return anything until the provider's signed
// attestation matches what it authorized. Authority comes from signatures and
// hashes — never from text a model emitted. This file walks that whole loop.
//
// Run it:  npm run example   (from typescript/)
import {
  buildManifest, // builds an UNSIGNED manifest with a correct content-addressed id
  signManifest, // Ed25519-signs it (signature covers JCS(manifest-without-signature))
  Ed25519Signer, // pluggable Ed25519 signer backed by an in-memory key
  argumentHash, // sha256(JCS(args)) — binds a call to its exact arguments (§8)
  hash, // sha256(JCS(value)) — used by the provider to attest its result
} from "@vcp/sdk";
import { invoke, type Provider } from "@vcp/gateway"; // the enforcing end-to-end flow
import { signAttestation } from "@vcp/gateway"; // a provider signs its result (§9)
import { DefaultPolicy } from "@vcp/gateway"; // a reference taint-aware policy authority
import type { ResultEnvelope } from "@vcp/sdk";

// Two independent keys: one identity owns the capability (signs the manifest AND
// its result attestations); the Gateway has its own key (mints grants, signs audit).
// Separating them is the point — the Gateway trusts the provider's signature, not its word.
const providerSigner = Ed25519Signer.generate();
const gatewaySigner = Ed25519Signer.generate();

// 1. Declare a capability as a CONTRACT. The contract (schemas, effects,
//    determinism, sandbox) is what gets hashed into the identity — summaries are
//    advisory and excluded. effects.class "read-only" means no approval prompt.
const unsigned = buildManifest({
  issuer: "did:web:example.com",
  provider: "example.fx",
  name: "fx.convert",
  version: "1.0.0",
  summary_for_user: "Convert an amount (in minor units, e.g. cents) between currencies at a fixed demo rate.",
  summary_for_model: "Pure read-only currency conversion on integer minor units. No side effects.",
  input_schema: {
    type: "object",
    additionalProperties: false,
    properties: { amount: { type: "number" }, from: { type: "string" }, to: { type: "string" } },
    required: ["amount", "from", "to"],
  },
  output_schema: { type: "object", properties: { converted: { type: "number" } }, required: ["converted"] },
  effects: { class: "read-only", external_side_effect: false }, // read-only ⇒ auto-allowed, no approval
  determinism: { class: "pure", supports_dry_run: false },
  sandbox: { filesystem: "none", network: [], secrets: [] },
});

// 2. Sign it. The capability_id is content-addressed: vcp:cap:<name>@sha256:<contract hash>.
//    Change ANY contract field after signing and the id no longer matches — that's the
//    anti-"rug pull" property. The Gateway recomputes this hash on every verify.
const manifest = await signManifest(unsigned, providerSigner);
console.log("capability_id:", manifest.capability.id);

// 3. A 3-line in-process provider: it recomputes argument_hash (rejecting tampered args),
//    produces a result, and signs an attestation binding capability+args+result together.
const provider: Provider = {
  publicKey: () => providerSigner.publicKey(),
  async invoke(a): Promise<ResultEnvelope> {
    if (argumentHash(a.arguments) !== a.argument_hash) throw new Error("ARGUMENT_HASH_MISMATCH"); // §8
    // The "real" work. VCP v0.1 canonical JSON is integer-only, so we keep the
    // result an integer: convert minor units (cents) at a fixed demo rate of 92%.
    const result = { converted: Math.round((a.arguments.amount as number) * 92 / 100) };
    return { result, attestation: await signAttestation(
      { capability_id: a.capability_id, argument_hash: a.argument_hash, result_hash: hash(result), effect_committed: false },
      providerSigner) };
  },
};

// 4. Run the full gateway loop: verify manifest → policy → mint single-use grant →
//    invoke provider → verify the signed attestation → audit. Fails closed at every step.
const outcome = await invoke(
  {
    subject: "user:alice",
    manifest,
    arguments: { amount: 100, from: "USD", to: "EUR" },
    plan_hash: "sha256:" + "0".repeat(64), // a real planner would hash its proposed plan (proposePlan)
    jkt: providerSigner.thumbprint(), // proof-of-possession: the grant is bound to this key
  },
  {
    manifestTrustedKey: providerSigner.publicKey(), // key the Gateway trusts signed the manifest
    trustedIssuers: ["did:web:example.com"],
    policy: new DefaultPolicy(),
    gatewaySigner,
    provider,
  },
);

// 5. The result is only here because every check passed and the attestation verified.
console.log("ok:", outcome.ok, "result:", outcome.result, "reason:", outcome.reason_code ?? "(allowed)");
console.log("audit events:", outcome.audit.map((e) => e.event).join(", "));
